### ISSUE TITLE
DeepSeek V4 Flash fails during CUDA graph profiling with B12X FP8 GEMM illegal address after 2026-08-15 image update

### ISSUE THREAD

**Issue by technigmaai** [eugr/spark-vllm-docker#349]
## Description

After updating `eugr/spark-vllm-b12x:latest` on August 15, 2026, the `deepseek-v4-flash-0731` recipe no longer starts on my dual DGX Spark/GX10 cluster.

The model loads successfully, but the engine crashes during piecewise CUDA graph memory profiling with `CUDA_ERROR_ILLEGAL_ADDRESS`.

Rolling back to the August 13 image fixes the problem without any recipe or system configuration changes.

## Environment

- Hardware: 2× DGX Spark/GX10, SM121 (`sm_121a`)
- Tensor parallel size: 2
- Model: `deepseek-ai/DeepSeek-V4-Flash-0731`
- Current failing image ID:

```text
sha256:62749cdc94f9bbe92b028d4f6607d838cf690f0ac8c862cd159000a6dc1fb2b3
```

- Current failing image creation date: `2026-08-15 04:30:27 +0200`
- Failing vLLM version:

```text
0.1.dev20003+gad848fc41.d20260815
```

- Working rollback image ID:

```text
sha256:a2df4abebff70246df85a5eeba15cd0490733fed76f516f36e0281587c1b15d7
```

- Working image creation date: `2026-08-13 19:50:54 +0200`

## Reproduction

```bash
./run-recipe.sh deepseek-v4-flash-0731
```

I also reproduced the same failure with a custom variant of the recipe. Both nodes use the same image ID, SSH connectivity succeeds, NCCL initializes, and all model weights load successfully.

Startup reaches:

```text
Profiling CUDA graph memory (PIECEWISE): 64% | 7/11
```

and then crashes.

## Relevant error

```text
cutlass.base_dsl.common.DSLCudaRuntimeError:
error: CUDA_ERROR_ILLEGAL_ADDRESS (error code: 700)
```

The traceback points to the B12X FP8 block-scaled GEMM used by a shared MoE expert:

```text
vllm/model_executor/layers/fused_moe/runner/shared_experts.py
vllm/models/deepseek_v4/nvidia/model.py
vllm/model_executor/kernels/linear/scaled_mm/b12x.py
b12x/gemm/_shared/block_fp8.py
b12x/_lib/dense_gemm.py
nvidia_cutlass_dsl/.../jit_executor.py
```

The failing operation is:

```text
torch.ops.b12x.block_fp8_linear_mxfp8_fused
```

The CUTLASS diagnostic reports:

```text
CUDA_TOOLKIT_PATH: not set
Target SM ARCH: unknown (unspecified)
Architecture: Blackwell (sm_121a)
Compatible SM archs: sm_121a
```

## Fresh compilation test

This does not appear to be caused by a stale vLLM compilation cache. On a fresh container/cache, the log showed that the graph was compiled and saved before failing:

```text
Cache the graph of compile range (1, 8192) for later use
Compiling a graph for compile range (1, 8192) takes 10.03 s
saved AOT compiled function
```

The newly compiled graph then failed during CUDA graph profiling.

There is also a warning immediately before compilation:

```text
torch.compile is turned on, but the model deepseek-ai/DeepSeek-V4-Flash-0731 does not support it
```

## Expected behavior

The engine should complete CUDA graph profiling and start the OpenAI-compatible API server, as it does with the August 13 image.

## Actual behavior

The B12X FP8/CUTLASS kernel encounters an illegal memory access during CUDA graph profiling. EngineCore fails to initialize and the API server exits.

## Confirmed workaround

Tagging and using the August 13 image restores normal operation:

```yaml
container: vllm-node-b12x:rollback-20260813
```

No other configuration changes are required.

This appears distinct from #348: that report reaches serving and later times out in `sample_tokens`, while this regression crashes during startup with a concrete B12X/CUTLASS `CUDA_ERROR_ILLEGAL_ADDRESS`.

Please let me know if the complete logs or a comparison of installed `b12x`, PyTorch, CUTLASS DSL, and vLLM revisions between the two images would help.

**Comment by pvutov** [comment]
Seeing the same  issue. Using `--enforce-eager` avoids the issue but is ofc not desirable

**Comment by technigmaai** [comment]
Thanks for the confirmations ;) at least it’s not only me ;)

**Comment by keeratita** [comment]
i can run the docker perfectly, however, i got malformed tool calls many time after 20260813 too.

**Comment by socode-marcelo** [comment]
Another data point on the same build, with a **different symptom** — the engine starts and serves fine here, but output corrupts under long generation.

Same hardware and recipe as the report: 2× DGX Spark (GX10), SM121, TP=2, `deepseek-v4-flash-0731`, no Ray.

```
failing : 0.1.dev20003+gad848fc41.d20260815     (image b72c96d38e98, built 2026-08-15)
working : 0.1.dev19023+g30038602b.d20260805     (image d43f15877df4)
```

Same vLLM commit as the original report, different image ID, so it looks like a rebuild of the same source.

### What we saw

No CUDA graph failure — startup was clean, graph capture completed, and it served normally for ~25 minutes. Then in an agentic session, after roughly 12k tokens of generation, the output degenerated to token-level garbage:

```
The intersection S‑‑‑‑‑ The‑ W‑ [‑‑‑‑‑‑ The‑-‑‑ The 745‑‑ The‑‑‑‑‑‑ e M‑‑ it
```

Runs of U+2011 non-breaking hyphens plus a cycling fragment. Worth distinguishing this from #341: that was coherent *sentence-level* looping (valid text in a bad order). This is invalid text.

Short generations were clean throughout — a cold-prefill garble probe passed 0/10 cold and 0/5 warm on this build, capped at 120 tokens. It only shows up deep into a decode, which is why it survived our pre-flight checks.

### The part that might be diagnostic

Measuring the same synthetic workload on both images, **prefill is unchanged and only decode regressed**:

- pure prefill, bucketed by concurrency, p50 identical on both builds (same chunk-steps per logging interval)
- single-stream decode down roughly 8–12% at 2k / 20k / 60k context depth

If this were generally slower kernels, prefill should have moved too. It didn't move at all. That asymmetry points at the speculative decode path rather than the shared compute path.

Speculative acceptance also looked lower on the new build, and degraded more at later draft positions — but I'd treat that as suggestive only. Acceptance is very sensitive to traffic mix, and the comparison available here was a small synthetic sample against a reference gathered from real traffic, so the two aren't properly comparable. Flagging the direction, not the magnitude.

Combined with the `--enforce-eager` workaround reported above (which disables CUDA graphs), a draft/graph interaction seems like a reasonable place to look.

### Rollback

Reverting to the `d20260805` image with no recipe change restored correct output. Re-ran the same long-generation probe on it: clean, 0/3 at 12k tokens.

One other difference between the builds, in case it's useful — with an identical `--kv-cache-memory-bytes 17179869184`:

```
d20260805 : GPU KV cache size 2,198,756 tokens   (2.10x)
d20260815 : GPU KV cache size 2,412,569 tokens   (2.30x)
```

Happy to run something specific on either image if it would help narrow it down.

**Comment by gabriellhl** [comment]
i am able to run with latest version also, but i can see garbled in the return

**Comment by keeratita** [comment]
> Another data point on the same build, with a **different symptom** — the engine starts and serves fine here, but output corrupts under long generation.
> 
> Same hardware and recipe as the report: 2× DGX Spark (GX10), SM121, TP=2, `deepseek-v4-flash-0731`, no Ray.
> 
> ```
> failing : 0.1.dev20003+gad848fc41.d20260815     (image b72c96d38e98, built 2026-08-15)
> working : 0.1.dev19023+g30038602b.d20260805     (image d43f15877df4)
> ```
> 
> Same vLLM commit as the original report, different image ID, so it looks like a rebuild of the same source.
> 
> ### What we saw
> No CUDA graph failure — startup was clean, graph capture completed, and it served normally for ~25 minutes. Then in an agentic session, after roughly 12k tokens of generation, the output degenerated to token-level garbage:
> 
> ```
> The intersection S‑‑‑‑‑ The‑ W‑ [‑‑‑‑‑‑ The‑-‑‑ The 745‑‑ The‑‑‑‑‑‑ e M‑‑ it
> ```
> 
> Runs of U+2011 non-breaking hyphens plus a cycling fragment. Worth distinguishing this from [#341](https://github.com/eugr/spark-vllm-docker/issues/341): that was coherent _sentence-level_ looping (valid text in a bad order). This is invalid text.
> 
> Short generations were clean throughout — a cold-prefill garble probe passed 0/10 cold and 0/5 warm on this build, capped at 120 tokens. It only shows up deep into a decode, which is why it survived our pre-flight checks.
> 
> ### The part that might be diagnostic
> Measuring the same synthetic workload on both images, **prefill is unchanged and only decode regressed**:
> 
> * pure prefill, bucketed by concurrency, p50 identical on both builds (same chunk-steps per logging interval)
> * single-stream decode down roughly 8–12% at 2k / 20k / 60k context depth
> 
> If this were generally slower kernels, prefill should have moved too. It didn't move at all. That asymmetry points at the speculative decode path rather than the shared compute path.
> 
> Speculative acceptance also looked lower on the new build, and degraded more at later draft positions — but I'd treat that as suggestive only. Acceptance is very sensitive to traffic mix, and the comparison available here was a small synthetic sample against a reference gathered from real traffic, so the two aren't properly comparable. Flagging the direction, not the magnitude.
> 
> Combined with the `--enforce-eager` workaround reported above (which disables CUDA graphs), a draft/graph interaction seems like a reasonable place to look.
> 
> ### Rollback
> Reverting to the `d20260805` image with no recipe change restored correct output. Re-ran the same long-generation probe on it: clean, 0/3 at 12k tokens.
> 
> One other difference between the builds, in case it's useful — with an identical `--kv-cache-memory-bytes 17179869184`:
> 
> ```
> d20260805 : GPU KV cache size 2,198,756 tokens   (2.10x)
> d20260815 : GPU KV cache size 2,412,569 tokens   (2.30x)
> ```
> 
> Happy to run something specific on either image if it would help narrow it down.

yep, i got the same thing. thanks for very nice details 👍

**Comment by keeratita** [comment]
it seems the issue has been fixed already as i tested from the latest image https://hub.docker.com/r/eugr/spark-vllm-b12x?tag=nightly-20260822, after running it in the real project for many hours.

<img width="157" height="131" alt="Image" src="https://github.com/user-attachments/assets/a66dc343-1574-4eb5-97ca-99269c0409a6" />

**Comment by maci0** [comment]
Different failure than ours (B12X CuteDSL GEMM IMA during PIECEWISE profiling vs host-staged PYNCCL). Related notes from a 2x GB10 0731 bring-up on stock v0.28.0rc2 + b12x==1.2.6, `FLASHINFER_MLA_SPARSE_DSV4`, DSpark k=5:

- spark2 earlyoom SIGTERM at MemAvailable <8% (~10 GiB), prefers `comm=vllm`. util 0.85 died (docker exit 0). util 0.81 stayed up with ~10.7 GiB available. SwapTotal is 0.
- Host-staged PYNCCL inside breakable CUDA graphs corrupts logits or dies in `cudaGraphLaunch`. `--enforce-eager` is coherent. PIECEWISE worked after eager-breaking TP all-reduce and cloning the buffer off the CUDA graph pool. FULL still bad.
- This recipe sets `VLLM_USE_BREAKABLE_CUDAGRAPH=0` and `FULL_AND_PIECEWISE`. Worth isolating graphs vs B12X GEMM capture as two bugs.

Not claiming this fixes `torch.ops.b12x.block_fp8_linear_mxfp8_fused`. Greedy France on our FLASHINFER path is `" Paris. The capital of Spain is Madrid..."`.

**Comment by keeratita** [comment]
so, is the issue part of b12x?
